Index: paramiko-1.7.7.1/paramiko/client.py
===================================================================
--- paramiko-1.7.7.1.orig/paramiko/client.py	2011-05-22 01:57:09.000000000 +0000
+++ paramiko-1.7.7.1/paramiko/client.py	2012-07-05 00:38:50.000000000 +0000
@@ -303,11 +303,7 @@
 
         server_key = t.get_remote_server_key()
         keytype = server_key.get_name()
-
-        if port == SSH_PORT:
-            server_hostkey_name = hostname
-        else:
-            server_hostkey_name = "[%s]:%d" % (hostname, port)
+        server_hostkey_name = hostname
         our_server_key = self._system_host_keys.get(server_hostkey_name, {}).get(keytype, None)
         if our_server_key is None:
             our_server_key = self._host_keys.get(server_hostkey_name, {}).get(keytype, None)
